### ISSUE TITLE
Flutter web : How to build in a single Flutter project a build for Chrome Extension and for Browser

### ISSUE THREAD

**Issue by redDwarf03** [flutter/flutter#106428]
Flutter allows to build app for browsers and for chrome extension BUT, the structure of index.html is not the same between these 2 builds.
Is it possible to build 2 apps in a single project ?

**Comment by huycozy** [comment]
Hi @redDwarf03, thanks for filing the issue.

Could you please describe the difference between two `index.html` files for Web and Extension? Which steps did you try to do? And then please update the issue description as detailed as possible.

I see that there is a tutorial [here](https://medium.com/flutter-community/building-a-chrome-extension-using-flutter-aeb100a6d6c) that builds an extension (contains modification steps for `index.html`) but I'm not sure is the same as yours or not.

**Comment by redDwarf03** [comment]
Hello. thanks for your answer. 
i saw this tutorial too and the author is right but my need is more like that.

I have a project i want to build in several web mode:
1) i build my project with ```flutter build web --web-renderer canvaskit --release``` to host my app on firebase 
2) i build my project with ```flutter build web --web-renderer html --csp``` to publish on chrome extension

### The need 1) requires : 

**index.html :**
```
<!DOCTYPE html>
<html>

<head>
  <!--
    If you are serving your web app in a path other than the root, change the
    href value below to reflect the base path you are serving from.

    The path provided below has to start and end with a slash "/" in order for
    it to work correctly.

    For more details:
    * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base

    This is a placeholder for base href that will be replaced by the value of
    the `--base-href` argument provided to `flutter build`.
  -->

  <meta charset="UTF-8">
  <meta content="IE=Edge" http-equiv="X-UA-Compatible">
  <meta name="description" content="">

  <!-- iOS meta tags & icons -->
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  <meta name="apple-mobile-web-app-title" content="">
  <link rel="apple-touch-icon" href="icons/Icon-192.png">

  <!-- Favicon -->
  <link rel="icon" type="image/png" href="favicon.png" />

  <title>Archethic Wallet</title>
  <link rel="manifest" href="manifest.json">

  <script>
    // The value below is injected by flutter build, do not touch.
    var serviceWorkerVersion = null;
  </script>

  <!-- This script adds the flutter initialization JS code -->
  <script src="flutter.js" defer></script>
  <script src="splash/splash.js"></script>
  <meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" name="viewport" />
  <link rel="stylesheet" type="text/css" href="splash/style.css">
  <meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" name="viewport" />
</head>

<body>
  <script>
    window.addEventListener('load', function (ev) {
      // Download main.dart.js
      _flutter.loader.loadEntrypoint({
        serviceWorker: {
          serviceWorkerVersion: serviceWorkerVersion,
        }
      }).then(function (engineInitializer) {
        return engineInitializer.initializeEngine();
      }).then(function (appRunner) {
        return appRunner.runApp();
      });
    });
  </script>
</body>

</html>
```

**manifest.json** :

```
{
  "name": "",
  "short_name": "",
  "start_url": ".",
  "display": "standalone",
  "background_color": "#000000",
  "theme_color": "#000000",
  "description": "",
  "orientation": "portrait-primary",
  "prefer_related_applications": false,
  "icons": [
      {
          "src": "icons/Icon-192.png",
          "sizes": "192x192",
          "type": "image/png"
      },
      {
          "src": "icons/Icon-512.png",
          "sizes": "512x512",
          "type": "image/png"
      },
      {
          "src": "icons/Icon-maskable-192.png",
          "sizes": "192x192",
          "type": "image/png",
          "purpose": "maskable"
      },
      {
          "src": "icons/Icon-maskable-512.png",
          "sizes": "512x512",
          "type": "image/png",
          "purpose": "maskable"
      }
  ]
}
```


### The need 2) requires :

**index.html :**

```
<!DOCTYPE html>
<html style="height: 600px; width: 350px">

<head>
  <!--
    If you are serving your web app in a path other than the root, change the
    href value below to reflect the base path you are serving from.

    The path provided below has to start and end with a slash "/" in order for
    it to work correctly.

    For more details:
    * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base

    This is a placeholder for base href that will be replaced by the value of
    the `--base-href` argument provided to `flutter build`.
  -->

  <meta charset="UTF-8">
  <meta content="IE=Edge" http-equiv="X-UA-Compatible">
  <meta name="description" content="">

  <!-- iOS meta tags & icons -->
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  <meta name="apple-mobile-web-app-title" content="">
  <link rel="apple-touch-icon" href="icons/Icon-192.png">

  <!-- Favicon -->
  <link rel="icon" type="image/png" href="favicon.png" />

  <title>Archethic Wallet</title>
  <link rel="manifest" href="manifest.json">

  <script src="splash/splash.js"></script>
  <meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" name="viewport" />
  <link rel="stylesheet" type="text/css" href="splash/style.css">
  <meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" name="viewport" />
</head>

<body>
  <script src="main.dart.js" type="application/javascript"></script>
</body>

</html>
```
 
**manifest.json** :
```{
    "name": "Archethic Wallet",
    "description": "Archethic Wallet",
    "version": "0.2.23",
    "content_security_policy": {
        "extension_pages": "script-src 'self' ; object-src 'self'"
    },
    "action": {
        "default_popup": "index.html",
        "default_icon": "icons/Icon-192.png"
    },
    "manifest_version": 3
}````

So I have to change index.html and manifest.json manually in web folder before launch build commands
Ideally, we could have 2 "web" folders like "web" and "chrome extension" and pass in the build command the folder we want to use to build the app

**Comment by huycozy** [comment]
@redDwarf03 
Thanks for your detailed response. I modified the sample code section for better readability. Labeling the issue for the team to come up with ideas/thoughts.

**Comment by ditman** [comment]
> Ideally, we could have 2 "web" folders like "web" and "chrome extension" and pass in the build command the folder we want to use to build the app

That sounds like a duplicate of:

* https://github.com/flutter/flutter/issues/78254#issuecomment-1139219323

(Will close as duplicate)

However, a quick question @redDwarf03: we eventually need to deprecate loading flutter directly with:

```html
  <script src="main.dart.js" type="application/javascript"></script>
```

Why is that required for the extension project? Why can't flutter.js be used in an extension?

**Comment by redDwarf03** [comment]
hmm.. i just followed how to create a chrome extension in this article: https://medium.com/flutter-community/building-a-chrome-extension-using-flutter-aeb100a6d6c
and the author precise in the end of the chapter one:
"Then only insert the following <script> tag inside the <body>:
<script src="main.dart.js" type="application/javascript"></script>"

Sadly i can't tell more about that...

and i agree with you, it seems to be like issue #78254

**Comment by ditman** [comment]
@redDwarf03 can you test in your extension if embedding with `flutter.js` works (similarly to what you use for firebase)?

The fact that flutter web works as a browser extension is a lucky "accident", not a supported use case (we don't have any testing, for example, to guarantee that Flutter web works as a extension, and that we don't break it in the future!)

**Comment by redDwarf03** [comment]
i added ``` <script src="flutter.js" defer></script>```in the ```<HEAD>``` section
and i removed ``` <script src="main.dart.js" type="application/javascript"></script>```

i hope that's correspond to your request.
The chrome extension doesn't work with this updates. i'm blocked on the splash image

```
<!DOCTYPE html>
<html style="height: 600px; width: 350px">

<head>
  <!--
    If you are serving your web app in a path other than the root, change the
    href value below to reflect the base path you are serving from.

    The path provided below has to start and end with a slash "/" in order for
    it to work correctly.

    For more details:
    * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base

    This is a placeholder for base href that will be replaced by the value of
    the `--base-href` argument provided to `flutter build`.
  -->

  <meta charset="UTF-8">
  <meta content="IE=Edge" http-equiv="X-UA-Compatible">
  <meta name="description" content="Archethic">

  <!-- iOS meta tags & icons -->
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  <meta name="apple-mobile-web-app-title" content="Archethic Wallet">
  <link rel="apple-touch-icon" href="icons/Icon-192.png">

  <!-- Favicon -->
  <link rel="icon" type="image/png" href="favicon.png" />

  <title>Archethic Wallet</title>
  <link rel="manifest" href="manifest.json">

  <script src="splash/splash.js"></script>
  <meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" name="viewport" />
  <link rel="stylesheet" type="text/css" href="splash/style.css">
  <meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" name="viewport" />

  <script src="flutter.js" defer></script>
</head>

<body>
  <picture id="splash">
    <source
      srcset="splash/img/light-1x.png 1x, splash/img/light-2x.png 2x, splash/img/light-3x.png 3x, splash/img/light-4x.png 4x"
      media="(prefers-color-scheme: light)">
    <source
      srcset="splash/img/dark-1x.png 1x, splash/img/dark-2x.png 2x, splash/img/dark-3x.png 3x, splash/img/dark-4x.png 4x"
      media="(prefers-color-scheme: dark)">
    <img class="center" aria-hidden="true" src="splash/img/light-1x.png" alt="" />
  </picture>
</body>

</html>
```

**Comment by ditman** [comment]
In addition to the `flutter.js` you now need to initialize flutter on a script of your own within the `body` tag, please add something similar to this as the last tag of your `body` element (after the `</picture>`):

```html
<script>
  _flutter.loader.loadEntrypoint().then(function(engineInitializer) {
    return engineInitializer.autoStart();
  });
</script>
```

(More details in the [Customizing web app initialization](https://docs.flutter.dev/development/platform-integration/web/initialization) flutter doc.)

**Comment by redDwarf03** [comment]
2 errors:
````
Refused to execute inline script because it violates the following Content Security Policy directive: "script-src 'self'". Either the 'unsafe-inline' keyword, a hash ('sha256-vWDPfccMQe9SKhb+AZDDfc7nhFqrWRcX1gniVD+Uw+Q='), or a nonce ('nonce-...') is required to enable inline execution.
```

```
Refused to execute inline script because it violates the following Content Security Policy directive: "script-src 'self' 'wasm-unsafe-eval'". Either the 'unsafe-inline' keyword, a hash ('sha256-vWDPfccMQe9SKhb+AZDDfc7nhFqrWRcX1gniVD+Uw+Q='), or a nonce ('nonce-...') is required to enable inline execution
```
<img width="629" alt="Capture d’écran 2022-07-07 à 23 42 47" src="https://user-images.githubusercontent.com/68340133/177876379-66edf5ab-382d-4df6-9625-0221624b69b6.png">

**Comment by ditman** [comment]
@redDwarf03 those might have to do with this: https://developer.chrome.com/docs/extensions/mv3/intro/mv3-migration/#content-security-policy but I'm not an expert with Chrome extensions, so I can't offer much help there :/

**Comment by redDwarf03** [comment]
i try to migrate manifest in v2 to v3

{
  "manifest_version": 2,
  "version": "0.0.1",
  "name": "chrome_extension",
  "description": "A new Flutter project.",
  "browser_action": {
    "default_popup": "index.html",
    "default_icon": {
      "16": "icons/icon-192.png",
      "32": "icons/icon-192.png",
      "48": "icons/icon-192.png",
      "128": "icons/icon-192.png"
    }
  },
  "content_security_policy": "script-src https://unpkg.com/canvaskit-wasm@0.30.0/bin/canvaskit.js;"
},
my new manifest doesn't work

{
    "manifest_version": 2,
    "version": "0.0.1",
    "name": "chrome_extension",
    "description": "A new Flutter project.",
    "browser_action": {
        "default_popup": "index.html",
        "default_icon": {
            "16": "icons/icon-192.png",
            "32": "icons/icon-192.png",
            "48": "icons/icon-192.png",
            "128": "icons/icon-192.png"
        }
    },
    "content_security_policy": {
        "extension_pages": "https://unpkg.com/canvaskit-wasm@0.30.0/bin/canvaskit.js;",
        "sandbox": "script-src 'self' ; object-src 'self'"
    }
}
Any idea ?

i generate the chrome extension : flutter build web --web-renderer html

**Comment by PatrickGeyer** [comment]
My app uses firebase, so it the firebase plugin loads a JS script from apis.google.com.
V3 does not allow this. But there's no way I can change how the plugin JS is loaded. Any suggestions?

**Comment by ditman** [comment]
> the firebase plugin loads a JS script from apis.google.com.
> V3 does not allow this. But there's no way I can change how the plugin JS is loaded. Any suggestions?

@PatrickGeyer this question would be much better suited for StackOverflow or other dev resources, and I'll mark it as "off-topic", however, some suggestions:

> there's no way I can change how the plugin JS is loaded.

There's a _(maybe not too well [documented](https://www.google.com/search?q=%22flutterfire_ignore_scripts%22))_ `package:firebase_core_web` feature that instructs firebase to **not automatically download their JS**, and instead use whatever is available in the index.html. From [the Dart doc](https://github.com/firebase/flutterfire/blob/ab39d09e1a24ad6cfba25b35b2435aef342d8e87/packages/firebase_core/firebase_core_web/lib/src/firebase_core_web.dart#L51-L60):

> #### `FirebaseCoreWeb` getter `_ignoredServiceScripts`
>
> Returns a list of services which won't be automatically injected on initialization.
>
> This is useful in cases where you wish to manually include the scripts (e.g. in countries where you must request the users permission to include Analytics).
> 
> You can do this by attaching an array of services to the window, e.g:
>
> ```
> window.flutterfire_ignore_scripts = ['analytics'];
> ```
>
> You must ensure the Firebase script is injected before using the service.

In order for the Firebase scripts to be available on the `index.html` of your extension, you probably need to bundle your `main.dart.js` with the Firebase scripts you need (from `npm`) and use this "bundle.js" output as the code for your extension. Some people have had success (in general, not explicitly with Flutter web) here:
  * [Group: `chromium-extensions`: Manifest v3 & Firebase](https://groups.google.com/a/chromium.org/g/chromium-extensions/c/xQmZLc8cu6Q)

I hope the above general info helps. If you need more details, you probably should hit the appropriate Stack Overflow tags.
